contracts/eosio.system/include/eosio.system/eosio.system.hpp

```diff
@@ -282,6 +284,43 @@ namespace eosiosystem {
       EOSLIB_SERIALIZE( producer_info2, (owner)(votepay_share)(last_votepay_share_update) )
    };
 
+   // finalizer_key_info stores information about a finalizer key.
+   struct [[eosio::table("finkeys"), eosio::contract("eosio.system")]] finalizer_key_info {
+      uint64_t          id;                   // automatically generated ID for the key in the table
+      name              finalizer_name;       // name of the finalizer owning the key
+      std::string       finalizer_key;        // finalizer key in base64url format
+      std::vector<char> finalizer_key_binary; // finalizer key in binary format in Affine little endian non-montgomery g1
+      checksum256       finalizer_key_hash;   // hash of finalizer key, cached to avoid re-computing
+
+      uint64_t    primary_key() const { return id; }
+      uint64_t    by_fin_name() const { return finalizer_name.value; }
+      checksum256 by_fin_key()  const { return finalizer_key_hash; }
+   };
+   typedef eosio::multi_index<
+      "finkeys"_n, finalizer_key_info,
+      indexed_by<"byfinname"_n, const_mem_fun<finalizer_key_info, uint64_t, &finalizer_key_info::by_fin_name>>,
+      indexed_by<"byfinkey"_n, const_mem_fun<finalizer_key_info, checksum256, &finalizer_key_info::by_fin_key>>
+   > finalizer_keys_table;
+
+   // finalizer_info stores information about a finalizer.
+   struct [[eosio::table("finalizers"), eosio::contract("eosio.system")]] finalizer_info {
+      name              finalizer_name;              // finalizer's name
+      uint64_t          active_finalizer_key_id;     // finalizer's active finalizer key's id in finalizer_keys_table, for fast finding key information
+      std::vector<char> active_finalizer_key_binary; // finalizer's active finalizer key's binary format in Affine little endian non-montgomery g1
+      uint32_t          finalizer_key_count = 0;     // number of finalizer keys registered by this finalizer
+
+      uint64_t primary_key() const { return finalizer_name.value; }
+   };
+   typedef eosio::multi_index< "finalizers"_n, finalizer_info > finalizers_table;
+
+   // last_fin_key_info stores information about a finalizer key used in last finalizer policy.
+   struct [[eosio::table("lastfinkeys"), eosio::contract("eosio.system")]] last_fin_key_info {
+      uint64_t          id;  // finalizer key's id in finalizer_keys_table
+
+      uint64_t primary_key() const { return id; }
+   };
+   typedef eosio::multi_index< "lastfinkeys"_n, last_fin_key_info > last_fin_keys_table;
```

Comment:
Thanks for your detailed analysis! I was initially worried about taking too much RAM to store finalizer_authorities. But that made the code more complicated (those kept and new finalizer IDs variables I had to introduce), not straightforward to reason, and miss potential corner cases.

Done.

Reply:
You don't have to use an actual `singleton`. I know there are problems with the way it is implemented (`static`) that hurt testing. You could instead use a table that only has one entry.

In fact, it may be better for the table that holds the last proposed finalizer set to not be a singleton the same way the other global singletons are handled. In the constructor of `system_contract`, we always read and deserialize the value of the singleton even if it is not going to be used by the action that will execute. This is an inefficient pattern that would be better if we stop following.

The table only needs to be read in the some cases; specifically when we need to call the `set_finalizers` host function or when a finalizer key is activated and the action needs to check if the previously active key was in the last proposed finalizer set.